### DISCUSSION TITLE
Tasmota-supported WiFi Thermostat Socket with external thermal sensor

### DISCUSSION THREAD

**Discussion by bam80** [arendst/Tasmota#20125] [Support request]
Hello, anyone has a sample of Tasmota-supported WiFi Thermostat Socket with external thermal sensor, something like SimPal-TY130 or similar?
https://www.simpal.cn/products/41-TY130%20Tuya%20WiFi%20Socket.html
![image](https://github.com/arendst/Tasmota/assets/233979/886e5d87-63a9-40d4-9775-da440ec51c6e)

I know there is Shelly Plus Plug S but I suppose it's internal thermal sensor doesn't quite suit for thermostat control (meaning the device will heat itself):
 https://templates.blakadder.com/shelly_plus_plug_S.html

**Comment by barbudor** [comment]
First time I see such kind of device, generally devices that support external sensors are DIN mountable (Sonoff THD) or Shellys with their add-on.

Alternatively if you are good at hacking you could get a Sonoff S26(R2) that are quite easy to open and should be able to find a connection inside on a free GPIO for a DS18x20 probe
If you go that way, beware not to choose a model which includes Power Monitoring as those plugs have hte ESP directly connected to the mains without isolation. 
The S26 doesn't have power monitoring so it qualify for such hack.

**Reply by sfromis** [reply]
You also have the Sonoff TH16 with a similar input for a sensor probe, but that's not a plug-type device.

**Reply by bam80** [reply]
Thanks, I considered DIN-mountable options but my current assumption is plug-type device could be preferable for thermostat control as otherwise I would need two controllers - the first one for temperature sensor in DIN and the other one is for actually smart plug (as I can't directly wire my thermostat to DIN-mountable device, easily).
So the idea was to combine both functions (thermal sensor and relay) in one device, for simplicity.

The other thing that comes to mind if my fears are actually justified - does my assumption make sense that Shelly Plus Plug S can heat itself under reasonable load, so it's internal sensor is not quite suitable for external thermostat control, or in normal situation it shouldn't get warm anyway so I have wrong assumption here?

**Reply by sfromis** [reply]
Indeed, internal temperature sensors in a plug are **not** for ambient temperature (I'd expect values significantly above), but only meant as a safety cutoff. For simplicity, I prefer having sensors separate, and then to have more data than only temperature.

**Reply by barbudor** [reply]
In the direction of separated sensors, I'm using wireless T&H sensor (BLE or Zigbee) with a gateway that publish the temperature over MQTT.
The heater is then driven by a relay-only device which receive the room temp from MQTT and manage the thermostat. 

That leave you completely free of the location of the sensor without any wires, and a free top-up with the small LCD on the sensor giving you the room temperature 😉

No need to have sensor and relay physically attached

**Reply by bam80** [reply]
So your sensors/LCDs are battery powered? How long do they last?

**Reply by barbudor** [reply]
More than a year, close to 2
Battery is a CR2032

I only have BLE ones. I would assume Zigbee would last at least the same.

**Reply by sfromis** [reply]
My room sensors is typically with sensors like BME280, SCD40 and TSL2591. And wired power, not at all focus on "interior design" ...

**Reply by bam80** [reply]
> More than a year, close to 2

That's quite a lot, but considering there will be several devices (~5), it still might be a problematic.

> My room sensors is typically with sensors like BME280, SCD40 and TSL2591. And wired power, not at all focus on "interior design" ...

Right now I'm considering finished devices, not much DIY hacking (yet).

**Reply by barbudor** [reply]
I have around 20 of Mija BLE sensors
They are cheap, good battery life. You can even monitor battery level and raise alarm weeks ahead of failure.
You can replace the original firmware by open source Pvvx firmware which remove the need for decryption (and that allow a lot of fine tuning if you want but I stick on default settings)

**Reply by bam80** [reply]
> In the direction of separated sensors, I'm using wireless T&H sensor (BLE or Zigbee) with a gateway that publish the temperature over MQTT.

I couldn't find any info if some cheap WiFi sensors, e.g.  [Tuya WIFI Temperature & Humidity Sensor (TH01)](https://www.youtube.com/watch?v=ZxI7jYJAG3Q), can be connected to Tasmota device?
It seems HA supports those, though.

**Reply by barbudor** [reply]
Wifi battery powered sensors are a bad design choice due to the higher power consumption of Wifi vs BLE/Zigbee.
If you want to run Wifi sensors, follow @sfromis way using a development board such as NodeMCU, ESP-DevKit with sensors modules such as BME280, HDC1080 or even environment sensors such as e-CO2, particule, etc ...
More over as Tuya has moved away from ESP to their own Wifi modules using chips that are not supported by Tasmota, you would likely end up needing to unsolder the Tuya wifi module and replace it with an ESP based one to support Tasmota.

Not a cheap and simple solution

Alternatively, those Xiaomi Mija sensors (search on Ali for [LYWSD03MMC](https://www.aliexpress.com/item/1005001626975979.html)) with an [ESP32 dev board](https://www.aliexpress.com/item/1005006154020092.html) running `tamota32-bluetooth` and there you are.
You just need to put the board into a small box and get an spare USB charger to power it. 
The ESP32 will publish the sensors on your MQTT broker
And your plug(s) will each get the measure and manage the thermostat.

As said earlier with a Shelly Plus Plug which has en ESP32solo1, you should be able to run Bluetooth on it, removing the need for a MQTT server if you don't want one. But that requires a self build tasmota configuration (although there are cloud based tools such as TasmoCompiler which makes that a breeze)

**Reply by sfromis** [reply]
Note that if you put dev board and sensors in a box, you still need to be careful about distance, physical placement and airflow inside the box. And do not rely on an offset calculation to adjust for the heat generated by the dev board, or any Wifi usage.

**Reply by bam80** [reply]
First of all, thank you both for supporting me!

After all, I decided to give it a go and take DIY sensor way, at least for the first prototype item.
I like simplicity of Wemos platform, so current plan is to buy something with USB-C and BLE (for further possible experiments with Xiaomi Mija sensors) from their Mini or Pico series ([C3](https://www.wemos.cc/en/latest/c3/index.html)/[S3](https://www.wemos.cc/en/latest/s3/s3_mini.html)?) and put some Environment Shield on it: 
https://www.wemos.cc/en/latest/d1_mini_shield/dht.html
https://www.wemos.cc/en/latest/d1_mini_shield/sht30.html

Could you recommend something that works out-of-the-box reliably with Tasmota?

**Reply by sfromis** [reply]
While that SHT30 shield should in principle work fine with Tasmota, mounting such a temperature shield stacked with an D1 Mini (or similar) is just a bad idea, as a "radiator" and temperature sensor are bad neighbors. Other than than, my sensor builds are similar, making sure to separate sensor and heating source (the ESP chip) by distance, and avoidance of heat conducting pathways and likely airflows. More generally, there are several I2C sensors working nicely with Tasmota, and the SHT30 series it not a bad option.

For a temperature sensor, the older ESP8266-based D1 Mini should work fine, but using a ESP32-C3/S2/S3 variant does give "more toys" like better opportunities for local-only programming. Beware that the older V1.0.0 variant of the C3 Mini has a bad antenna design, and it may be better to get original branded boards instead of clones. I actually have all 4 chip variants of the D1 Mini form factor in operation. Ignoring price differences, my preference is the S3 Mini.

And a physical build not having the ESP chip heat the temperature sensor.

**Reply by bam80** [reply]
Thank you very much @sfromis,
noted your precautions about the temperature shield placement near MCU.
Having that, it comes to mind Shelly H&T sensor has exactly such design, rather it's sensor and MCU are placed on the same plane:
https://www.youtube.com/watch?v=6ppjGBKePq0

Anyway, the SHT30 seem has I2C connector so can be separated, and also I think I could try to use Pin Header Board Spacer to make it higher in the stack.

**Reply by sfromis** [reply]
As a battery-operated device with much lower power usage, sleeping most of the time, the Shelly H&T has a **lot** less risk of heating issues. Thus not comparable to an always-on device. While it is possible to make a similar build and using Tasmota deep sleep, the D1 Mini board is not suitable for this, at least without modification.

**Reply by bam80** [reply]
Agree. Just a note they also have [USB power supply for Shelly H&T](https://www.shelly.com/en/products/shop/usb-power-supply-for-shelly-h-and-t-white), not sure if it makes it "always on" device, too.

**Reply by sfromis** [reply]
Probably not, just the convenience of not having to switch batteries.

**Reply by bam80** [reply]
> While it is possible to make a similar build and using Tasmota deep sleep, the D1 Mini board is not suitable for this, at least without modification.

What modification do you mean, can I read somewhere about it?
Does it applies to C3/S3 also or they sleep properly?
Do I need to make custom build of Tasmota or just follow steps here:
https://tasmota.github.io/docs/DeepSleep/

> As a battery-operated device with much lower power usage, sleeping most of the time, the Shelly H&T has a lot less risk of heating issues.

I really hope to achieve the same with Tasmota, regardless if my device is battery-powered or not.
If I'm holding out false hopes, please let me know.

**Reply by barbudor** [reply]
The comment applies to most boards with an on-board usb/serial adapter has it's drawing current even when the esp asleep
As of today best ESP boards for battery operation are based on ESP32-C3 with direct USB connection (known as CDC) without usb/serial on board
Some of these boards may already include a battery circuitry to allow charging
Without battery charger: LuatOS C3 Core, Seedstudio Xiao C3, TenStar Robot SuperMini C3
With battery charger: looks at Liligo

ESP8266 is now quite old and not really optimized for battery operations. More recent ESP fits better.
But chips with wifi will always use more juice than BLE or Zigbee.

**Reply by barbudor** [reply]
There's that video from Andreas Spiess on how he modified a mini D1.
I did similar hack with partial success
https://youtu.be/rfPwOtoGO4E

**Reply by sfromis** [reply]
Good candidates for battery operation are also the boards by Unexpected Maker, as those are specifically tuned for efficient battery operation.

**Reply by bam80** [reply]
Thanks
>  With battery charger: looks at Liligo

~You mean LOLIN?~
Oh, I see those Liligo, will look closely, thanks.

Anyway,
I'm considering Wemos LOLIN C3 Pico but didn't quite get yet if Tasmota would be able to calculate and show battery fuel gauge for me from it's V+ BAT wired to ADC pin, as there is no I2C fuel gauge on this board.

**Reply by sfromis** [reply]
Wiring battery directly to an analog input pin can only be recommended if you want to see how long (short) time until fried. Using a voltage divider, you could scale it to a specs-compliant range, but assuming lithium rechargeable batteries, you cannot make a good "fuel gauge" based on the voltage alone, as the voltage curve from a freshly charged battery starts high, is flat most of the time, until getting near empty. Of course, for many purposes it is "good enough" with an an alarm for "better switch/recharge battery now".

**Reply by bam80** [reply]
Yes, the board was designed that way, see the VBAT jumper:
![image](https://github.com/arendst/Tasmota/assets/233979/5d6ff5a3-9d85-4b30-bff1-205ca1eb6b8a)
It also has the voltage divider:
![image](https://github.com/arendst/Tasmota/assets/233979/92dd11f1-e3c1-4f92-b623-e163e058d74a)
https://www.wemos.cc/en/latest/_static/files/sch_c3_pico_v1.0.0.pdf

So I'm trying to understand if Tasmota has the facility to help me get useful info from that AD_BAT pin.

**Reply by sfromis** [reply]
As that specific board does have that voltage divider for a safe voltage range, instead of direct battery connection, you should get an analog value on IO3 which is half the battery voltage (the voltage divider has same resistance on both sides). You may have to solder the VBAT "arrow" for connection.

**Reply by bam80** [reply]
Yes, understand that, but as you mentioned yourself, that voltage alone won't give me a clue about the full gauge.
So I expect maybe there is some algorithm to deduce it or something that Tasmota could support for such boards.

**Reply by sfromis** [reply]
If you know the average discharge rate and range, and when last full charge was, the calculation to get an estimate of mAh spent and left is reasonably simple, and any deviation found when it gets low could be factored into the calculations.

Or simpler: Assuming stable power draw, you could make one full run until discharged, and collect readings and how long time. If your result is that it can last 5 days, and "now" is 3 days since last full charge, it follows that a reasonable estimate is 2 days more, or a bit less as a safety margin. And using the voltages, you should have a good indication of when it was fully charged.

**Reply by barbudor** [reply]
Tasmota supports MAX17043 fuel gauge 
https://www.aliexpress.com/item/32665716677.html

**Reply by bam80** [reply]
> Of course, for many purposes it is "good enough" with an an alarm for "better switch/recharge battery now".

On a second thought, probably that will be enough.
I'll just push VBAT voltage values over MQTT and display the gauge in IOT MQTT or some other UI.
Important thing I need to organize alarm somehow if the voltage is close to 3.2 V. Push notification will suit but I don't know if IOT MQTT allows me to do it.

The other thing to make sure if cutoff battery below 3.2 V is implemented somewhere on the board or in battery's controller itself.

**Comment by Wintespe** [comment]
I dare to doubt that the tiny transformer of the S26 and similar devices really maintains a series separation while maintaining all insulation distances. Creepage distances, safety distances etc. must be maintained in the transformer. I would not recommend anyone to rely on this. The manufacturer does not normally give a guarantee for mains isolation.

**Reply by bam80** [reply]
This seems has common roots with what @barbudor talked about:
> If you go that way, beware not to choose a model which includes Power Monitoring as those plugs have hte ESP directly connected to the mains without isolation.

Sorry I'm newbie so don't quite follow what kind of insulation (and "ESP" connection?) you both are talking about, could you clarify? Maybe with some external references, thanks.

**Reply by barbudor** [reply]
Normaly those ESP devices have in the box a power supply which is not much different than your traditional USB charger for your phone.
Between the mains and the low voltage circuit there is a transformer which provide insulation and voltage reduction. However modern power supply design combine a circuit before the transformer (on mains side) and another after the transformer which allow them to work with much smaller transformer than what was existing before.
The ESP processor is connected exclusively to the low voltage side of the circuit. So if you connect a sensor to the ESP GPIOs through long wires going outside of the original enclosure, there are normally no risks to get shocked. No more than when touching your mobile while it is charging on the USB charger.

I believe @Wintespe is pointing that still those devices are not designed to be hacked and by doing so you are still taking a risk. If you solder your wire at a wrong location or mistakenly make contact with a wire that belong to the "hot" side of the circuit, you'll get shocked.

Devices with power measurment are different. Because the power measurment has to be done on the mains side, the ESP must be in direct contact with the part of the circuit which perform the measurment. Which means that the ESP is connected to the mains.
Normally this is not a problem as long as the original enclosure remains closed an no-one tries to hack into.
But if you connect a sensor to the ES through wires going out of the enclose, then your sensor is also connected to the mains. 
🔥 

I hopes this clarifies

I never attached sensors specifically to a S26 plug but I did attached AM2103 and HDC1080 Temp & Hum sensors to Sonoff Basic R1.
I did checked before that the GPIOs were absolutely isolated from mains.

**Reply by sfromis** [reply]
In case of bad hardware failures, devices without good separation just might get mains voltage on the low voltage side, and then any wires out of the device could get mains voltage, which risks both electrical shocks and starting fires. While I'd not call it a big risk, insurance companies might be "reluctant" to pay if such a tinkered device turns out to have started a fire.

One of the reasons why I like having the sensors on separate devices with dev boards is that any stupid errors from my side has a much lower risk of getting dangerous.

**Reply by barbudor** [reply]
The 4th first minutes of this presentation explains why modern power supplies based on switching mode have drastically reduced the size of the transformer (some basic understanding of electronics is required): https://www.youtube.com/watch?v=-ZYt589LauQ

One important point is that there is still a transformer => insulation between the 2 sides

Which still hide annother point : on those circuit, the mains voltage is rectified from AC to DC. Which means that a small part of the circuit carries DC 230V whivch is actually WAY MORE DANGEROUS than AC 230V

So yes, even while the ESP and its GPIOs is isolated, a misplaced or wandering wire could be dangerous.

If you want to be safe: Get some BLE or Zigbee sensors with any plug.
Normally you need a gateway with those sensors that would push the measures to MQTT:
Sensor =BLE/ZB=> Gateway =MQTT=> Broker =MQTT=> Plug

But with a Shelly Plus Plug which is using an ESP32, you shuld be able to use BLE directly on the Plug, so you could "pair" a BLE sensor with the Plug without the need of any additional infrastructure. Still the plug could report to MQTT Temp, humiduty and battery level

**Reply by barbudor** [reply]
Regarding BLE sensors battery life, here is a snapshot for the last 70 days (max I keep in influxdb) where you can see that despite very erratic values, this is going down quite slow:
![image](https://github.com/arendst/Tasmota/assets/2996042/c4e78882-ecb1-47fb-956d-87d9e9ff7c74)

**Reply by bam80** [reply]
> But with a Shelly Plus Plug which is using an ESP32, you shuld be able to use BLE directly on the Plug, so you could "pair" a BLE sensor with the Plug without the need of any additional infrastructure. Still the plug could report to MQTT Temp, humiduty and battery level

Thanks, is it documented somewhere? Does it compiled-in to the Tasmota pre-compiled binaries?

**Reply by sfromis** [reply]
For Bluetooth support, you need a binary including this feature.
I'm seeing LYWSD03 temperatures using precompiled `tasmota32-bluetooth.bin`.

**Reply by bam80** [reply]
> I hopes this clarifies

@barbudor it more or less does, thanks.

I found a video which sheds some light to the problem of hacking Sonoff POW devices (still, to some extent for me):
"#99 Measuring Power, hacked Sonoff POW, MQTT, Adafruit.io and Home Automation"
https://youtu.be/hv_FpBqWFrg?si=4vCq67_bJ_IWjiov

Still, it seems to me we should document it somewhere more clearly, maybe on the dev's Wiki pages?
https://templates.blakadder.com/sonoff_POWR320D.html
Or, if that applies to all the Power Metering devices - in the Hardware Preparation[~](https://tasmota.github.io/docs/Getting-Started/#hardware-preparation) section?
It already has some precaution:
>  DO NOT CONNECT DEVICES TO MAINS AC POWER WHILE THE COVER IS OPEN AND CIRCUIT BOARD IS EXPOSED!!!

but it doesn't emphasize Power Metering devices peculiarities, whereas the usual way to power up for many is to exactly CONNECT DEVICES TO MAINS AC POWER (BTW, how to solve it in this case?)

**Reply by bam80** [reply]
> Devices with power measurment are different. Because the power measurment has to be done on the mains side, the ESP must be in direct contact with the part of the circuit which perform the measurment. Which means that the ESP is connected to the mains.
Normally this is not a problem as long as the original enclosure remains closed an no-one tries to hack into.
But if you connect a sensor to the ES through wires going out of the enclose, then your sensor is also connected to the mains.
🔥

Still, in my understanding if the wires to the sensor are properly isolated and the sensor with it's PCB are enclosed in case which doesn't fall apart upon falling on the floor, such hack could be considered relatively safe? If you know what your are doing, of course.

I'll use a separate device for a sensor of the prototype, but later on I consider return to this hacky solution maybe.

I wonder if I could use this sensor for the hack, it looks pretty isolated:
![image](https://github.com/arendst/Tasmota/assets/233979/ec9822ac-59b8-422d-bbe7-2d4460359d92)

https://www.seeedstudio.com/DHT22-wired-AM2311-temperature-humidity-sensor-Arduino-sensor-p-4195.html

**Reply by sfromis** [reply]
It is **not** safe with thin wires if there is any risk of seeing mains power on those wires. The image you show does not count as "pretty isolated", just flimsy thin wires without much isolation. Just do not connect anything to the inside of energy monitoring devices.
